File: Backend/functions/src/moderationMeshContracts.ts
export type SafetyCapabilityKind =
    | "contentSafety"
    | "childSafety"
    | "scamFraud"
    | "accountIntegrity"
    | "harassmentBrigading"
    | "linkSafety"
    | "creatorProtection"
    | "communityHealth"
    | "crisis"
    | "imageSafety"
    | "liveVoice"
    | "liveVideo"
    | "deepfakeProvenance";

export type SafetyBackendKind = "managed" | "nvidia" | "appleOnDevice" | "stub";

export type SafetySignalLevel = "none" | "low" | "medium" | "high" | "critical";

export type AdvisoryAction =
    | "allow"
    | "nudge"
    | "holdForHumanReview"
    | "routeToCrisisWorkflow"
    | "restrictDistribution"
    | "knownCSAMProviderHardBlock";

export interface SafetyMeshFlags {
    moderation_mesh_enabled: boolean;
    managed_safety_backends_enabled: boolean;
    nvidia_safety_backends_enabled: boolean;
    text_safety_agents_enabled: boolean;
    image_safety_pipeline_enabled: boolean;
    trust_graph_internal_scores_enabled: boolean;
    creator_protection_agent_enabled: boolean;
    community_health_agent_enabled: boolean;
    crisis_agent_enabled: boolean;
    live_voice_moderation_enabled: boolean;
    live_video_moderation_enabled: boolean;
    c2pa_provenance_enabled: boolean;
    csam_hash_scan_enabled: boolean;
}

export const defaultSafetyMeshFlags: SafetyMeshFlags = Object.freeze({
    moderation_mesh_enabled: false,
    managed_safety_backends_enabled: false,
    nvidia_safety_backends_enabled: false,
    text_safety_agents_enabled: false,
    image_safety_pipeline_enabled: false,
    trust_graph_internal_scores_enabled: false,
    creator_protection_agent_enabled: false,
    community_health_agent_enabled: false,
    crisis_agent_enabled: false,
    live_voice_moderation_enabled: false,
    live_video_moderation_enabled: false,
    c2pa_provenance_enabled: false,
    csam_hash_scan_enabled: false,
});

export interface CSAMComplianceGate {
    espNcmecRegistrationComplete: boolean;
    hashProviderContractSigned: boolean;
    writtenLegalSignoffComplete: boolean;
    nonEngineerReviewComplete: boolean;
}

export const closedCSAMComplianceGate: CSAMComplianceGate = Object.freeze({
    espNcmecRegistrationComplete: false,
    hashProviderContractSigned: false,
    writtenLegalSignoffComplete: false,
    nonEngineerReviewComplete: false,
});

export interface SafetyCapabilityPayload {
    contentId?: string;
    contentType: "text" | "image" | "link" | "account" | "behavior" | "provenance" | "audio" | "video";
    storageRef?: string;
    url?: string;
    text?: string;
    metadata?: Record<string, unknown>;
}

export interface EvidenceRef {
    refId: string;
    kind: "policy" | "modelSignal" | "behavioralSignal" | "providerDecision" | "userReport" | "provenance";
    summary: string;
}

export interface AdvisoryVerdict {
    verdictId: string;
    capability: SafetyCapabilityKind;
    backend: SafetyBackendKind;
    signal: string;
    level: SafetySignalLevel;
    confidence: number;
    recommendedAction: AdvisoryAction;
    evidenceRefs: EvidenceRef[];
    requiresHumanReview: boolean;
    autonomousActionPermitted: boolean;
    createdAt: string;
}

export interface SafetyCapabilityProvider {
    capability: SafetyCapabilityKind;
    backend: SafetyBackendKind;
    providerVersion: string;
    enabled: boolean;
}

export interface GuardianSignalRecord {
    signalId: string;
    socialContextId: string;
    actorUid?: string;
    targetUid?: string;
    surfaceKind: string;
    capability: SafetyCapabilityKind;
    verdict: AdvisoryVerdict;
    createdAt: string;
    expiresAt?: string;
}

export interface HumanReviewQueueItem {
    itemId: string;
    signalId: string;
    socialContextId: string;
    priority: SafetySignalLevel;
    status: "pending" | "assigned" | "resolved" | "dismissed";
    capability: SafetyCapabilityKind;
    recommendedAction: AdvisoryAction;
    evidenceRefs: EvidenceRef[];
    createdAt: string;
    assignedToUid?: string;
}

export const SafetyMeshInvariants = Object.freeze({
    advisoryOnlyAgents: true,
    nvidiaIsOptionalPostLaunch: true,
    liveVoiceVideoContractsOnly: true,
    deepfakeUsesC2PAProvenanceOnly: true,
    publicTrustScoresAllowed: false,
    countryBasedRiskScoringAllowed: false,
    e2eeDeferredForServerSideScanning: true,
});

export function isCSAMHashScanAllowed(flags: SafetyMeshFlags, gate: CSAMComplianceGate): boolean {
    return flags.csam_hash_scan_enabled === true
        && gate.espNcmecRegistrationComplete === true
        && gate.hashProviderContractSigned === true
        && gate.writtenLegalSignoffComplete === true
        && gate.nonEngineerReviewComplete === true;
}

export function isCapabilityEnabled(kind: SafetyCapabilityKind, flags: SafetyMeshFlags): boolean {
    if (!flags.moderation_mesh_enabled) {
        return false;
    }
    switch (kind) {
    case "contentSafety":
    case "childSafety":
    case "scamFraud":
    case "accountIntegrity":
    case "harassmentBrigading":
    case "linkSafety":
        return flags.text_safety_agents_enabled && flags.managed_safety_backends_enabled;
    case "creatorProtection":
        return flags.creator_protection_agent_enabled && flags.managed_safety_backends_enabled;
    case "communityHealth":
        return flags.community_health_agent_enabled;
    case "crisis":
        return flags.crisis_agent_enabled && flags.managed_safety_backends_enabled;
    case "imageSafety":
        return flags.image_safety_pipeline_enabled && flags.managed_safety_backends_enabled;
    case "deepfakeProvenance":
        return flags.c2pa_provenance_enabled;
    case "liveVoice":
        return false;
    case "liveVideo":
        return false;
    }
}

export function normalizeAdvisoryVerdict(verdict: AdvisoryVerdict): AdvisoryVerdict {
    const autonomousActionPermitted = verdict.recommendedAction === "knownCSAMProviderHardBlock";
    return {
        ...verdict,
        confidence: Math.max(0, Math.min(1, verdict.confidence)),
        requiresHumanReview: verdict.requiresHumanReview || verdict.level === "high" || verdict.level === "critical",
        autonomousActionPermitted,
    };
}

export function requiresHumanReview(verdict: AdvisoryVerdict): boolean {
    return normalizeAdvisoryVerdict(verdict).requiresHumanReview;
}

File: AMENAPP/ActionThreads/ModerationMeshContracts.swift
import Foundation

// Swift mirror of Backend/functions/src/moderationMeshContracts.ts.
// Keep this file contract-only: providers and routing are implemented in later waves.

enum SafetyCapabilityKind: String, Codable, CaseIterable {
    case contentSafety
    case childSafety
    case scamFraud
    case accountIntegrity
    case harassmentBrigading
    case linkSafety
    case creatorProtection
    case communityHealth
    case crisis
    case imageSafety
    case liveVoice
    case liveVideo
    case deepfakeProvenance
}

enum SafetyBackendKind: String, Codable, CaseIterable {
    case managed
    case nvidia
    case appleOnDevice
    case stub
}

enum SafetySignalLevel: String, Codable, CaseIterable, Comparable {
    case none
    case low
    case medium
    case high
    case critical

    private var rank: Int {
        switch self {
        case .none: return 0
        case .low: return 1
        case .medium: return 2
        case .high: return 3
        case .critical: return 4
        }
    }

    static func < (lhs: SafetySignalLevel, rhs: SafetySignalLevel) -> Bool {
        lhs.rank < rhs.rank
    }
}

enum AdvisoryAction: String, Codable, CaseIterable {
    case allow
    case nudge
    case holdForHumanReview
    case routeToCrisisWorkflow
    case restrictDistribution
    case knownCSAMProviderHardBlock
}

struct SafetyMeshFlags: Codable, Equatable {
    let moderationMeshEnabled: Bool
    let managedSafetyBackendsEnabled: Bool
    let nvidiaSafetyBackendsEnabled: Bool
    let textSafetyAgentsEnabled: Bool
    let imageSafetyPipelineEnabled: Bool
    let trustGraphInternalScoresEnabled: Bool
    let creatorProtectionAgentEnabled: Bool
    let communityHealthAgentEnabled: Bool
    let crisisAgentEnabled: Bool
    let liveVoiceModerationEnabled: Bool
    let liveVideoModerationEnabled: Bool
    let c2paProvenanceEnabled: Bool
    let csamHashScanEnabled: Bool

    static let disabled = SafetyMeshFlags(
        moderationMeshEnabled: false,
        managedSafetyBackendsEnabled: false,
        nvidiaSafetyBackendsEnabled: false,
        textSafetyAgentsEnabled: false,
        imageSafetyPipelineEnabled: false,
        trustGraphInternalScoresEnabled: false,
        creatorProtectionAgentEnabled: false,
        communityHealthAgentEnabled: false,
        crisisAgentEnabled: false,
        liveVoiceModerationEnabled: false,
        liveVideoModerationEnabled: false,
        c2paProvenanceEnabled: false,
        csamHashScanEnabled: false
    )
}

struct CSAMComplianceGate: Codable, Equatable {
    let espNcmecRegistrationComplete: Bool
    let hashProviderContractSigned: Bool
    let writtenLegalSignoffComplete: Bool
    let nonEngineerReviewComplete: Bool

    static let closed = CSAMComplianceGate(
        espNcmecRegistrationComplete: false,
        hashProviderContractSigned: false,
        writtenLegalSignoffComplete: false,
        nonEngineerReviewComplete: false
    )

    var isCleared: Bool {
        espNcmecRegistrationComplete
            && hashProviderContractSigned
            && writtenLegalSignoffComplete
            && nonEngineerReviewComplete
    }
}

struct SafetyCapabilityPayload: Codable, Equatable {
    enum ContentType: String, Codable, CaseIterable {
        case text
        case image
        case link
        case account
        case behavior
        case provenance
        case audio
        case video
    }

    let contentId: String?
    let contentType: ContentType
    let storageRef: String?
    let url: URL?
    let text: String?
    let metadata: [String: String]
}

struct EvidenceRef: Codable, Identifiable, Equatable {
    enum Kind: String, Codable, CaseIterable {
        case policy
        case modelSignal
        case behavioralSignal
        case providerDecision
        case userReport
        case provenance
    }

    let id: String
    let kind: Kind
    let summary: String

    init(id: String = UUID().uuidString, kind: Kind, summary: String) {
        self.id = id
        self.kind = kind
        self.summary = summary
    }
}

struct AdvisoryVerdict: Codable, Identifiable, Equatable {
    let id: String
    let capability: SafetyCapabilityKind
    let backend: SafetyBackendKind
    let signal: String
    let level: SafetySignalLevel
    let confidence: Double
    let recommendedAction: AdvisoryAction
    let evidenceRefs: [EvidenceRef]
    let requiresHumanReview: Bool
    let autonomousActionPermitted: Bool
    let createdAt: Date

    init(
        id: String = UUID().uuidString,
        capability: SafetyCapabilityKind,
        backend: SafetyBackendKind,
        signal: String,
        level: SafetySignalLevel,
        confidence: Double,
        recommendedAction: AdvisoryAction,
        evidenceRefs: [EvidenceRef],
        requiresHumanReview: Bool,
        autonomousActionPermitted: Bool = false,
        createdAt: Date = Date()
    ) {
        self.id = id
        self.capability = capability
        self.backend = backend
        self.signal = signal
        self.level = level
        self.confidence = min(max(confidence, 0), 1)
        self.recommendedAction = recommendedAction
        self.evidenceRefs = evidenceRefs
        self.requiresHumanReview = requiresHumanReview || level >= .high
        self.autonomousActionPermitted = recommendedAction == .knownCSAMProviderHardBlock && autonomousActionPermitted
        self.createdAt = createdAt
    }
}

struct SafetyCapabilityProviderDescriptor: Codable, Equatable {
    let capability: SafetyCapabilityKind
    let backend: SafetyBackendKind
    let providerVersion: String
    let enabled: Bool
}

protocol SafetyCapabilityProvider {
    var descriptor: SafetyCapabilityProviderDescriptor { get }

    func evaluate(
        context: SocialContext,
        payload: SafetyCapabilityPayload
    ) async throws -> AdvisoryVerdict
}

struct GuardianSignalRecord: Codable, Identifiable, Equatable {
    let id: String
    let socialContextId: String
    let actorUid: String?
    let targetUid: String?
    let surfaceKind: String
    let capability: SafetyCapabilityKind
    let verdict: AdvisoryVerdict
    let createdAt: Date
    let expiresAt: Date?
}

struct HumanReviewQueueItem: Codable, Identifiable, Equatable {
    enum Status: String, Codable, CaseIterable {
        case pending
        case assigned
        case resolved
        case dismissed
    }

    let id: String
    let signalId: String
    let socialContextId: String
    let priority: SafetySignalLevel
    let status: Status
    let capability: SafetyCapabilityKind
    let recommendedAction: AdvisoryAction
    let evidenceRefs: [EvidenceRef]
    let createdAt: Date
    let assignedToUid: String?
}

enum SafetyMeshInvariants {
    static let advisoryOnlyAgents = true
    static let nvidiaIsOptionalPostLaunch = true
    static let liveVoiceVideoContractsOnly = true
    static let deepfakeUsesC2PAProvenanceOnly = true
    static let publicTrustScoresAllowed = false
    static let countryBasedRiskScoringAllowed = false
    static let e2eeDeferredForServerSideScanning = true

    static func isCSAMHashScanAllowed(flags: SafetyMeshFlags, gate: CSAMComplianceGate) -> Bool {
        flags.csamHashScanEnabled && gate.isCleared
    }

    static func isCapabilityEnabled(_ kind: SafetyCapabilityKind, flags: SafetyMeshFlags) -> Bool {
        guard flags.moderationMeshEnabled else { return false }

        switch kind {
        case .contentSafety, .childSafety, .scamFraud, .accountIntegrity, .harassmentBrigading, .linkSafety:
            return flags.textSafetyAgentsEnabled && flags.managedSafetyBackendsEnabled
        case .creatorProtection:
            return flags.creatorProtectionAgentEnabled && flags.managedSafetyBackendsEnabled
        case .communityHealth:
            return flags.communityHealthAgentEnabled
        case .crisis:
            return flags.crisisAgentEnabled && flags.managedSafetyBackendsEnabled
        case .imageSafety:
            return flags.imageSafetyPipelineEnabled && flags.managedSafetyBackendsEnabled
        case .deepfakeProvenance:
            return flags.c2paProvenanceEnabled
        case .liveVoice, .liveVideo:
            return false
        }
    }
}
